lib/lookdevXUsd/UsdMxVersionUpgrade.cpp

```diff
@@ -0,0 +1,1097 @@
+        {_tokens->boolean, SdfValueTypeNames->Bool},
+        {_tokens->filename, SdfValueTypeNames->Asset},
+        {_tokens->string, SdfValueTypeNames->String},
+    };
+
+
+
+namespace {
+
+Ufe::Path _getMaterialPath(const Ufe::Path& materialElementPath)
+{
+    if (materialElementPath.empty()) {
+        return {};
+    }
+
+    auto sceneItem = Ufe::Hierarchy::createItem(materialElementPath);
+
+    while (sceneItem && sceneItem->nodeType() != _tokens->Material) {
+        auto hierarchy = Ufe::Hierarchy::hierarchy(sceneItem);
+        sceneItem = hierarchy->parent();
+    }
+    if (!sceneItem) {
+        return {};
+    }
+    return sceneItem->path();
+}
+
+std::optional<std::string> _isLegacyMaterial(const UsdShadeMaterial& material)
+{
+    auto mxSurfaceOutput = material.GetSurfaceOutput(_tokens->mtlx);
+    if (!mxSurfaceOutput || !mxSurfaceOutput.HasConnectedSource())
+    {
+        // No MaterialX shading in this material.
+        return std::nullopt;
+    }
+    // Fetch the version from the MaterialXConfigAPI schema:
+    auto materialXVersion = std::string{"1.38"}; // Default to 1.38 if not authored.
+
+    const auto configAPI = UsdMtlxMaterialXConfigAPI(material);
+    if (configAPI)
+    {
+        const auto versionAttr = configAPI.GetConfigMtlxVersionAttr();
+        if (versionAttr)
+        {
+            // Got an authored version.
+            versionAttr.Get(&materialXVersion);
+        }
+    }
+
+    if (materialXVersion != _tokens->currentMxVersion.GetString())
+    {
+        return _tokens->legacyVersionPrefix.GetString() + materialXVersion;
+    }
+    
+    return std::nullopt;
+}
+
+std::pair<std::string, std::string> _splitNumericalSuffix(const std::string& srcName)
+{
+    // Compiled regular expression to find a numerical suffix to a path component.
+    // It searches for any number of characters followed by a single non-numeric,
+    // then one or more digits at end of string.
+    static const auto reSuffix = std::regex("(.*[^0-9])([0-9]+)$");
+    std::smatch base_match;
+ 
+    if (std::regex_match(srcName, base_match, reSuffix)) {
+        return std::make_pair(base_match[1].str(), base_match[2].str());
+    }
+    return std::make_pair(srcName, "");
+}
+
+TfToken _uniqueName(const std::set<std::string>& existingNames, const std::string& srcName)
+{
+    auto [base, suffixStr] = _splitNumericalSuffix(srcName);
+    size_t suffix = 1;
+    size_t suffixLen = 1;
+    if (!suffixStr.empty()) {
+        suffix = std::stoi(suffixStr) + 1;
+        suffixLen = suffixStr.size();
+    }
+
+    // Create a suffix string from the number keeping the same number of digits as
+    // numerical suffix from input srcName (padding with 0's if needed).
+    const auto buildName = [](auto suffix, auto suffixLen, auto& suffixStr, const auto& base) -> std::string {
+        suffixStr = std::to_string(suffix);
+        suffixStr = std::string(suffixLen - std::min(suffixLen, suffixStr.size()), '0') + suffixStr;
```

Comment:
Surprised and saddened that USD Tf stringUtil does not contain a padding function. Oh well. At least the code protects itself against negative with that min.